git-worktree-switcher: init at 0.2.4

jiriks74 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jiriks74 commented:
I think that all everything should now work as expected. I'll look into doing tests as that's something I am not familiar with.
But I think that the rest can be reviewed.

mateusauler commented:
Awesome! I'm willing to co-maintain the package, if you'd like.

jiriks74 commented:
> Awesome! I'm willing to co-maintain the package, if you'd like.

I'll gladly add you if you're in the maintainers file.

jiriks74 commented:
@mateusauler

I didn't find you in [the maintainers file](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix).
I can add you if you fork my changes, follow [the README](https://github.com/NixOS/nixpkgs/tree/master/maintainers#individual-maintainer-list) for adding maintainers and make a PR. I **cannot** add you as when you add a GPG signature to a maintainer the commit **has to** be signed with the key added.

jiriks74 commented:
I just added this package and module to my NixOS config. I had to make a modification to the module though:

```diff
diff --git a/modules/nixos/git-worktree-switcher.nix b/modules/nixos/git-worktree-switcher.nix
index f1462ab..490c3a2 100644
--- a/modules/nixos/git-worktree-switcher.nix
+++ b/modules/nixos/git-worktree-switcher.nix
@@ -2,6 +2,7 @@
   config,
   pkgs,
   lib,
+  ...
 }:
 
 let
```

or I'd get

```nix
       error: function 'anonymous lambda' called with unexpected argument 'outputs'
       at /nix/store/qc8rd1yhqd06yjfcnpigkfpnfgmyz2dy-source/modules/nixos/git-worktree-switcher.nix:1:1:
            1| {
             | ^
            2|   config,
```

So it seems like the missing `...` creates an issue with flakes.

---

Is the no `...` requirement only for packages or for modules too? (I understood from the docs to avoid it if possible)

I'll be looking through other modules and adding it back if I see `...` being commonly used.

mateusauler commented:
Whoops, forgot to format the maintainers section in the package file.

jiriks74 commented:
> Whoops, forgot to format the maintainers section in the package file.

I'll take care of it. I'll also add back the `...` since I misunderstood the docs and it's kinda needed for the module to work.

mateusauler commented:
> I'll take care of it.

Thanks.

> I'll also add back the `...` since I misunderstood the docs and it's kinda needed for the module to work.

I'll test it on my config as well and report any problems other than this one.

FliegendeWurst commented:
Please make sure your commit messages follow the guidelines.

mateusauler commented:
I believe it would be best not to include github's PR merge commit. As far as I can tell, just running `git rebase -i HEAD~2` and dropping it doesn't break anything.

jiriks74 commented:
> I believe it would be best not to include github's PR merge commit. As far as I can tell, just running git rebase -i HEAD~2 and dropping it doesn't break anything.

Makes sense. Should've thought of that before XD.
I droped the last two commits (the merge commit didn't show in rebase) and then I merge it again using git itself.

Hopefully everything's ready now.

mateusauler commented:
I noticed the module isn't in `nixos/modules/module-list.nix` and doesn't work in my config probably because of that.

jiriks74 commented:
> I noticed the module isn't in nixos/modules/module-list.nix and doesn't work in my config probably because of that.

Should be fixed now.

jiriks74 commented:
Well now I can either `git rebase` and break @mateusauler's commit again or I can `git merge` and have the merge commit there.

Which one you think would be better for this situation?

jiriks74 commented:
Or I can remove @mateusauler's commit altogether, finish this, and add him as a comaintainer once everything is done.

mateusauler commented:
Feel free to break my commit or remove it :). If you end up removing it, I'll create a PR, after this one is merged, and add myself back.
I'll also test everything later today and report back here.

jiriks74 commented:
> After all's working we can add it back or I'll create a PR to add myself as a maintainer.

I think that this is the best course of action too. I'll be looking out for your PR.

I'll look into testing this probably this evening. I have all the changes in my private Flake for my systems so the code itself works, it only needs to be added `nixpkgs` in a way that works.

jiriks74 committed:
nixos/git-worktree-switcher: init git-worktree-switcher

This module sets up shells so that they work with
[git-worktree-switcher](https://github.com/mateusauler/git-worktree-switcher)

pluiedev commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,36 @@
+  prg = config.programs;
+  cfg = prg.git-worktree-switcher;
+
+  bashInitScript = ''
```
This seems awfully redundant. I think we can just have one init script with the shell name substituted in

jiriks74 replied:
Fixed in 3d30c8d9340903897037bdf0d8f4eed08e56fa12

pluiedev commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,60 @@
+  jq,
+}:
+
+stdenv.mkDerivation rec {
```
Consider using finalAttrs

jiriks74 replied:
I'll look into this. I've basically modified the `thefuck` package to get this as I am not 100% familiar with all the options.

jiriks74 replied:
Fixed in 7c2dda4abd2b5cef40276badf4f57dd391bc209f

pluiedev commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,60 @@
+    installShellCompletion --fish completions/wt.fish
+  '';
+
+  meta = with lib; {
```
Avoid `with lib;`

pluiedev replied:
> What should I use instead then? I haven't found much documentation about this.

See FliegendeWurst's proposed changes

jiriks74 replied:
Fixed in 7c2dda4abd2b5cef40276badf4f57dd391bc209f

pluiedev reviewed (approved):
Overall LGTM, only some minor nitpicks - thanks for contributing to Nixpkgs!

pluiedev commented on maintainers/maintainer-list.nix:
```diff
@@ -10357,6 +10357,18 @@
     githubId = 1476865;
     name = "jigglycrumb";
   };
+  jiriks74 = {
```
Oh and this should be a separate commit

jiriks74 replied:
Fixed in b489ea0f4e1904db70c37bfbfe481fa2cc84c9f3

FliegendeWurst commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,60 @@
+    installShellCompletion --fish completions/wt.fish
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/mateusauler/git-worktree-switcher";
+    description = "Switch between git worktrees with speed.";
+    license = licenses.mit;
+    platforms = platforms.all;
+    maintainers = with maintainers; [ jiriks74 ];
```
```suggestion
  meta = {
    homepage = "https://github.com/mateusauler/git-worktree-switcher";
    description = "Switch between git worktrees with speed.";
    license = lib.licenses.mit;
    platforms = lib.platforms.all;
    maintainers = with lib.maintainers; [ jiriks74 ];
```

jiriks74 committed:
maintainers: Add jiriks74

jiriks74 committed:
git-worktree-switcher: init at 0.2.4

[`git worktree` documentation](https://git-scm.com/docs/git-worktree)
[`git-worktree-switcher` repository](https://github.com/mateusauler/git-worktree-switcher)

From the project's description:

> Switch between git worktrees with speed.

It's a wrapper script around `git worktree` making some tasks,
like switching between worktrees, easier.

> [!Note]
> `git worktree` is a little known feature and adding QoL things can help
it's adoption as it's a considerable upgrade against the `git stash`
workflow.

jiriks74 commented:
I've incorporated all the changes. The package builds as of now. I'll test the functionality shortly.

jiriks74 commented:
I just tested the package with `nixpkgs` set to my local repo in my OS config. Worked without issues.

mateusauler commented:
I've also tested it and everything seems fine. I found a bug with the script itself, which I'll fix later.

xanderio commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -201,6 +201,8 @@
 
 - [Glances](https://github.com/nicolargo/glances), an open-source system cross-platform monitoring tool. Available as [services.glances](option.html#opt-services.glances).
 
+- [git-worktree-switcher](https://github.com/mateusauler/git-worktree-switcher), switch between git worktrees with speed. Available as [programs.git-worktree-switcher](#opt-programs.git-worktree-switcher.enable)
```
24.11 has been release by now, could you please move this to the 25.05 release notes

xanderio commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,60 @@
+
+  patches = [
+    ./disable-update.patch # Disable update and auto update functionality
+    ./add_list_option.patch # Add `list` option to be more intuitive (`git worktree list`)
```
I don't think we should add functionality to a program via a patch in nixpkgs, could you send this upstream?

jiriks74 replied:
We already started a discussion on this. I agree that it shouldn't be here.

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+    environment.systemPackages = with pkgs; [ git-worktree-switcher ];
+
+    programs.bash.interactiveShellInit = initScript;
+    programs.zsh.interactiveShellInit = lib.mkIf prg.zsh.enable initScript;
```
```suggestion
    programs.zsh.interactiveShellInit = lib.optionalString config.progams.zsh.enable (initScript "zsh");
```

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+  prg = config.programs;
+  cfg = prg.git-worktree-switcher;
+
+  initScript = ''
+    eval "$(command ${pkgs.git-worktree-switcher}/bin/wt init $(basename $SHELL))"
```
```suggestion
  initScript =
    shell:
    if (shell != "fish") then
      ''
        eval $(${getExe pkgs.git-worktree-switcher} init ${shell})
      ''
    else
      ''
        ${getExe pkgs.git-worktree-switcher} init ${shell} | source
      '';```

jiriks74 replied:
May I ask why would you want use 2 different scripts when we can just use `eval $(${getExe pkgs.git-worktree-switcher} init ${shell})` for everything?

xanderio replied:
As fish is not a POSIX shell [`eval`](https://fishshell.com/docs/current/cmds/eval.html) isn't the intended way to load code into the shell, while `eval` may work in this case, let's please stick to piping into [`source`](https://fishshell.com/docs/current/cmds/source.html).

jiriks74 replied:
Fair enough. I just used what upstream docs said as I'm not familiar with `fish`. I'll look into implementing this fix on my laptop shortly.

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = with pkgs; [ git-worktree-switcher ];
+
+    programs.bash.interactiveShellInit = initScript;
```
```suggestion
    programs.bash.interactiveShellInit = initScript "bash";
```

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+
+    programs.bash.interactiveShellInit = initScript;
+    programs.zsh.interactiveShellInit = lib.mkIf prg.zsh.enable initScript;
+    programs.fish.interactiveShellInit = lib.mkIf prg.fish.enable initScript;
```
```suggestion
    programs.fish.interactiveShellInit = lib.optionalString config.programs.fish.enable (initScript "fish");
```

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+}:
+
+let
+  prg = config.programs;
```
```suggestion

```

xanderio commented on nixos/modules/programs/git-worktree-switcher.nix:
```diff
@@ -0,0 +1,30 @@
+
+let
+  prg = config.programs;
+  cfg = prg.git-worktree-switcher;
```
```suggestion
  cfg = config.programs.git-worktree-switcher;
```

xanderio reviewed (changes requested):
Thanks for your contribution!

The major thing that doesn't look right to me is the fish config part. I've suggested some changes that should make it work. Could you please tests them :)

mateusauler commented on pkgs/by-name/gi/git-worktree-switcher/disable-update.patch:
```diff
@@ -0,0 +1,80 @@
+ 	;;
+ *)
+-	auto_check_update
+ 	directory=$(git worktree list --porcelain | sed -n '/'"${arg:-.}"'/{s/^worktree\s*//p;q}')
```
```suggestion
 	directory=$(git worktree list --porcelain 2> /dev/null | sed -n '/'"${arg:-.}"'/{s/^worktree\s*//p;q}')
```

mateusauler commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,59 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "git-worktree-switcher";
+  version = "0.2.3";
```
```suggestion
  version = "0.2.4";
```

mateusauler commented on pkgs/by-name/gi/git-worktree-switcher/package.nix:
```diff
@@ -0,0 +1,59 @@
+    owner = "mateusauler";
+    repo = "git-worktree-switcher";
+    rev = "refs/tags/${finalAttrs.version}-fork";
+    hash = "sha256-fjY+ieUI7m/4gb68/1a+90aIc9iMyQortVd7oHN4STw=";
```
```suggestion
    hash = "sha256-N+bDsLEUM6FWhyliUav2n5hwMa5EEuVPoIK+Cja0DxA=";
```

mateusauler reviewed (commented):
@jiriks74 FYI, I just released v0.2.4.

jiriks74 commented:
Thanks for the notice @mateusauler.

I've simplified the patch to removing only the update calls, not the functions themselves (less potential breakages in the future) and updated the package to `0.2.4`.

jiriks74 commented:
@mateusauler You can make a pr and add yourself as a mainteiner now.
